Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/datadonation/analytics/storage/DefaultLastAnalyticsSubmissionLogger.kt

```diff
@@ -0,0 +1,87 @@
+package de.rki.coronawarnapp.datadonation.analytics.storage
+
+import android.content.Context
+import com.google.gson.Gson
+import com.google.gson.JsonParseException
+import com.google.gson.TypeAdapter
+import com.google.gson.stream.JsonReader
+import com.google.gson.stream.JsonWriter
+import de.rki.coronawarnapp.server.protocols.internal.ppdd.PpaData
+import de.rki.coronawarnapp.util.coroutine.DispatcherProvider
+import de.rki.coronawarnapp.util.di.AppContext
+import de.rki.coronawarnapp.util.serialization.BaseGson
+import de.rki.coronawarnapp.util.serialization.adapter.InstantAdapter
+import de.rki.coronawarnapp.util.serialization.fromJson
+import de.rki.coronawarnapp.util.serialization.toJson
+import kotlinx.coroutines.withContext
+import okio.ByteString.Companion.decodeBase64
+import okio.ByteString.Companion.toByteString
+import org.joda.time.Instant
+import org.json.JSONObject
+import timber.log.Timber
+import java.io.File
+import javax.inject.Inject
+
+class DefaultLastAnalyticsSubmissionLogger @Inject constructor(
+    @AppContext private val context: Context,
+    private val dispatcherProvider: DispatcherProvider,
+    @BaseGson private val baseGson: Gson
+) : LastAnalyticsSubmissionLogger {
+    private val analyticsDir = File(context.cacheDir, "analytics_storage")
+    private val analyticsFile = File(analyticsDir, "analytics.bin")
+
+    private val gson by lazy {
+        baseGson.newBuilder()
+            .registerTypeAdapter(Instant::class.java, InstantAdapter())
+            .registerTypeAdapter(PpaData.PPADataAndroid::class.java, PPADataAndroidAdapter())
+            .create()
+    }
+
+    override suspend fun storeAnalyticsData(analyticsProto: PpaData.PPADataAndroid) =
+        withContext(dispatcherProvider.IO) {
+            if (!analyticsDir.exists()) {
+                analyticsDir.mkdirs()
+            }
+
+            val dataObject = LastAnalyticsSubmission(
+                timestamp = Instant.now(),
+                ppaDataAndroid = analyticsProto
+            )
+
+            try {
+                gson.toJson(dataObject, analyticsFile)
+            } catch (e: Exception) {
+                Timber.e(e, "Failed to store analytics data.")
+            }
+        }
+
+    override suspend fun getLastAnalyticsData(): LastAnalyticsSubmission? = withContext(dispatcherProvider.IO) {
+        try {
+            gson.fromJson(analyticsFile)
```

Comment:
Based on issues we had with the exposure detection which also stores json to file:
We should prevent empty files from causing invalid data classes being created by GSON (examples in `ExposureDetectionTrackerStorageTest`).

Despite not being allowed GSON can create data classes with null fields that shouldn't be null (unsafe init), which the JVM will only notice once accessed.

TL;DR:  `.also { requireNotNull(it.timestamp) }`